### PR TITLE
Fix: prevent bound parameters being `[null]` when no parameters are given to `Predicate#expression()`

### PR DESCRIPTION
**Pull request by Ocramius**
Around `laminas/laminas-db:2.10.1`, a regression was introduced, in which calling `Laminas\Db\Sql\Predicate#expression("an_expression()")` led to crashes like following in downstream consumers:

```
Laminas\Db\Sql\Exception\RuntimeException: The number of replacements in the expression does not match the number of parameters

vendor/laminas/laminas-db/src/Sql/Expression.php:151
vendor/laminas/laminas-db/src/Sql/Predicate/PredicateSet.php:178
vendor/laminas/laminas-db/src/Sql/Predicate/PredicateSet.php:178
vendor/laminas/laminas-db/src/Sql/Predicate/PredicateSet.php:178
vendor/laminas/laminas-db/src/Sql/AbstractSql.php:129
vendor/laminas/laminas-db/src/Sql/Select.php:633
```

This was because predicates were initialized with an `array{null}` by default, when expressions like `$sql->where->expression("some_expression()")` were used.

The usage of `$sql->where->expression("some_expression()", "foo")` remains unchanged with this patch.

This fix targets `2.15.x`, and attempts to make predicates safe to use when no parameters have been given.

While an existing test has indeed been changed, this shouldn't have any effect for downstream consumers, since `Predicate#expression(string)` didn't work (so far) anyway, due to the kind of crash highlighted above.

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | yes
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

### CODE REVIEW

**Comment by Ocramius** [comment]
Note: yes, I targeted `2.15.x` because I'm abusing my maintainer privileges: need this for a paying customer, and will handle merge-up myself, once I have a review :D

File: test/unit/Sql/Predicate/PredicateTest.php [hunk 1]
```diff
@@ -276,4 +281,59 @@ public function testLiteral()
...
+    }
+
+    public function testWillBindSqlParametersToExpressionsWithGivenParameter(): void
+    {
+        $where = new Predicate();
+
+        $where->expression('some_expression(?)', null);
+
+        self::assertSame(
+            'SELECT "a_table".* FROM "a_table" WHERE (some_expression(\'\'))',
```
Comment (Ocramius) [inline test/unit/Sql/Predicate/PredicateTest.php:304]: Note: not interpolating `NULL` here, because that's what the Sql92 platform dictates.

File: src/Sql/Predicate/Predicate.php [hunk 2]
```diff
@@ -250,7 +250,7 @@ public function notLike($identifier, $notLike)
     public function expression($expression, $parameters = null)
     {
         $this->addPredicate(
-            new Expression($expression, $parameters),
+            new Expression($expression, func_num_args() > 1 ? $parameters : []),
```
Comment (internalsystemerror) [inline src/Sql/Predicate/Predicate.php:253]: If I invoke this like: `$predicate->expression('someExpression()', null)` will `func_num_args` be 2 in that case? And will it still break?

If it broke and before and breaks still:
a) do we need to add a test to ensure we get `[null]` in that case?
b) does that also need fixing?

Comment (Ocramius) [reply src/Sql/Predicate/Predicate.php:253]: > will `func_num_args()` be 2 in that case? 

Yes

 > do we need to add a test to ensure we get `[null]` in that case?

One of the tests added to this patch ensures that `[null]` is preserved

 > does that also need fixing?

No, I'd say that somebody using `->expression($string, $parameter)` is doing so intentionally, and we should leave it alone :)

**Review by internalsystemerror — APPROVED** [review]
LGTM 🚢

File: test/unit/Sql/Predicate/PredicateTest.php (hunk 1 above)
Comment (weierophinney) [reply test/unit/Sql/Predicate/PredicateTest.php:304]: That was my question.

I guess the bigger question I have is that the value is called `$parameters`, and type-hinted as an array. Passing `null` feels like you are saying "omit the parameters", versus "cast to a string".

The difficulty here, of course, is that the value was being passed verbatim to the `Expression` constructor, which allows `null`, and this would normally be cast as a string. I'm honestly not sure what the right answer is here; the previous behavior feels more like a bug than anything.

Comment (Ocramius) [reply test/unit/Sql/Predicate/PredicateTest.php:304]: Agreed: the problem here is that `null` being passed in is certainly an existing use-case, which we can't change at this stage.

I also had to scratch my eyes when I saw `null` being converted to `''`, but that's already the current use-case.
